### ISSUE TITLE
Otter randomly giving stats not recorded

### ISSUE THREAD

**Issue by andreped** [andreped/super-ml-pets#32]
During deployment, when buying the otter, it will give +2 attack +1 defence to a random member in the team. This is not recorded in the team.

**Comment by GoldExplosion** [comment]
For me, it gives +1 attack +1 defense to a random member in the team

before buying:
```
CURRENT INFO:
--------------
PACK:  StandardPack
TURN:  1
LIVES: 10
WINS:  0
GOLD:  7

CURRENT TEAM:
--------------
0: < Slot pet-beaver 2-2 none 1-0 >
    1: < Slot EMPTY >
    2: < Slot EMPTY >
    3: < Slot EMPTY >
    4: < Slot EMPTY >

CURRENT SHOP:
--------------
0: < ShopSlot-pet not-frozen 3-gold pet-duck 1-3 none 1-0 >
    1: < ShopSlot-pet not-frozen 3-gold pet-otter 1-2 none 1-0 >
    2: < ShopSlot-food not-frozen 3-gold food-honey 0-0 status-honey-bee >
    >
```
after buying:

```
CURRENT INFO:
--------------
PACK:  StandardPack
TURN:  1
LIVES: 10
WINS:  0
GOLD:  4

CURRENT TEAM:
--------------
0: < Slot pet-beaver 3-3 none 1-0 >
    1: < Slot pet-otter 1-2 none 1-0 >
    2: < Slot EMPTY >
    3: < Slot EMPTY >
    4: < Slot EMPTY >

CURRENT SHOP:
--------------
0: < ShopSlot-pet not-frozen 3-gold pet-duck 1-3 none 1-0 >
    1: < ShopSlot-food not-frozen 3-gold food-honey 0-0 status-honey-bee >
    >
```
This is because sapai hasn't been updated. Is this the issue or is the effect not recorded at all?

**Comment by andreped** [comment]
For training this is perfectly fine. The problem has to do with deployment with the actual game. We just have to catch which animal the otter gave the stats point.

**Comment by GoldExplosion** [comment]
oh I understand, the random pet the otter gave stats to in the actual game and in sapai is different. 
To find out which pet the game gave stats to will require an extension of the computer vision system so that it detects the pets and its stats in the team. (This is especially challenging because Pets may have effects such as honey on them which causes the confidence in detection  by single template matching to drop). 
One solution is to just detect the stats(health and attack) of the Pet in nth_slot and not the Pet itself(since we know what it is). This way we can know which Pet received the stat buff.

**Comment by andreped** [comment]
> This is especially challenging

Yes, this is absolutely not trivial. There are also other pets that have similiar complex behaviour. 

I think for now, I would not spend a lot of time on that. Now I think is the time to improve the reward system and RL algorithm, as I believe the deployment method is working fine enough.

It would also be great to support freezing of objects, but currently `sapai-gym` does not support freezing items:
https://github.com/alexdriedger/sapai-gym/blob/master/sapai_gym/SuperAutoPetsEnv.py#L294

We would also need to handle that in the machine vision system somehow. For instance the template matching method might fail if an animal is frozen. More tuning to be made.


> One solution is to just detect the stats(health and attack) of the Pet in nth_slot and not the Pet itself(since we know what it is). This way we can know which Pet received the stat buff.

I like that idea. Seems a lot more seemless. If you are interested, you can also contribute to the machine vision system. Up to you what you prefer working on :)

**Comment by GoldExplosion** [comment]
Yes, I am interested in contributing to the machine vision system. I will start when I finish what I am working on currently.

**Comment by andreped** [comment]
> Yes, I am interested in contributing to the machine vision system. I will start once I finish what I am working on currently.

Looking forward to it :)

**Comment by GoldExplosion** [comment]
There is a recent update to Super Auto Pets that allows you to set how you visualise the Food on a Pets. We can set the visualization of Food on Pets to be static. 
![image](https://user-images.githubusercontent.com/42666533/189033425-2e90be81-79d3-49f2-8827-de9cab535d97.png)

Now we can get state information about the current team. With this, we can switch between human player and AI player freely (previously this was not possible because there is no way for the AI to know what team was currently in the Team Slot). 

We can also check if the action made by the AI has been correctly implemented.

**Comment by andreped** [comment]
@GoldExplosion Oh!! That's great!! And as you said, this can easily be added through template matching.

But the machine vision system seems to know quite well which animals are in the item shop. It knowns the initial state and tracks all actions where it updates the team. I believe the team and team order should be available in the engine, at all times, or was there something else you were thinking about?

**Comment by andreped** [comment]
> restart the engine in the middle of the game

Yes, then I agree. For now, I have not taken that scenario into account, but that could easily be done, as you mentioned. However, we would have to update the current team and statuses to the engine. Should be possible.

However, improving the reward system is more urgent. I have no time this week to work on this project, but should have more time next week.

It is nice to build up a database, sure. However, it would be great if we didn't need "real games" to make a good AI ;) Right now, the AI does not seem to be that intelligent - I blame the reward system being too simple.

**Comment by GoldExplosion** [comment]
> However, improving the reward system is more urgent. I have no time this week to work on this project, but should have more time next week.

ok got it. Take your time. 

> It is nice to build up a database, sure. However, it would be great if we didn't need "real games" to make a good AI ;) Right now, the AI does not seem to be that intelligent - I blame the reward system being too simple.

I see. makes sense.
